### ISSUE TITLE
Upgrading to gradle 8.0.2 causes deadlock

### ISSUE THREAD

**Issue by NikolayMetchev** [gradle/gradle#24229]
We have attempted to upgrade from 7.6 to 8.0.2 and are seeing random deadlocks/hanging
Here is a jstack of the JVM in deadlock:
[jstack.txt](https://github.com/gradle/gradle/files/10921536/jstack.txt)

### Expected Behavior
Gradle to not deadlock and build our project as it does with version 7.6

### Current Behavior
deadlock/hang

### Context
Upgrade our build from 7.6 to 8.0.2

### Steps to Reproduce 
Unfortunately I don't have an easy way to reproduce this issue. I can create more jstack examples if you think that will help.

### Your Environment
Gradle 8.0.2
Gradle Enterprise Plugin 3.12.4
Java 17.0.6 (temurin)
MacOS Ventura
Linux( Docker in CI)

### Deadlocked threads

```
"Execution worker Thread 7" #468 prio=5 os_prio=31 cpu=507.48ms elapsed=117.19s tid=0x0000000140354c00 nid=0x1600b waiting on condition  [0x00000002b9771000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17.0.6/Native Method)
	- parking to wait for  <0x00000007932164a8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@17.0.6/LockSupport.java:341)
...
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@17.0.6/AbstractQueuedSynchronizer.java:1623)
	at org.gradle.internal.operations.DefaultBuildOperationQueue.lambda$waitForWorkToComplete$0(DefaultBuildOperationQueue.java:141)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$$Lambda$708/0x0000000800661f18.run(Unknown Source)
	at org.gradle.internal.Factories$1.create(Factories.java:31)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:321)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:304)
	at org.gradle.internal.work.DefaultWorkerLeaseService.blocking(DefaultWorkerLeaseService.java:210)
	at org.gradle.internal.operations.DefaultBuildOperationQueue.waitForWorkToComplete(DefaultBuildOperationQueue.java:135)
	at org.gradle.internal.operations.DefaultBuildOperationQueue.waitForCompletion(DefaultBuildOperationQueue.java:106)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.executeInParallel(DefaultBuildOperationExecutor.java:143)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.runAll(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.runAll(DefaultBuildOperationExecutor.java:97)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.verification.ChecksumAndSignatureVerificationOverride.verifyConcurrently(ChecksumAndSignatureVerificationOverride.java:129)
...
	at kotlinx.kover.appliers.KoverProjectApplierKt$engineProvider$1.call(KoverProjectApplier.kt)
	at org.gradle.api.internal.provider.DefaultProvider.calculateOwnValue(DefaultProvider.java:72)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.getOrNull(AbstractMinimalProvider.java:98)
	at org.gradle.internal.properties.bean.AbstractNestedRuntimeBeanNode.unpackProvider(AbstractNestedRuntimeBeanNode.java:83)
	at org.gradle.internal.properties.bean.AbstractNestedRuntimeBeanNode.visitProperties(AbstractNestedRuntimeBeanNode.java:60)
	at org.gradle.internal.properties.bean.NestedRuntimeBeanNode.visitNode(NestedRuntimeBeanNode.java:39)
	at org.gradle.internal.properties.bean.DefaultPropertyWalker.visitProperties(DefaultPropertyWalker.java:41)
	at org.gradle.api.internal.tasks.TaskPropertyUtils.visitAnnotatedProperties(TaskPropertyUtils.java:58)
	at org.gradle.api.internal.tasks.TaskPropertyUtils.visitProperties(TaskPropertyUtils.java:45)
	at org.gradle.api.internal.tasks.TaskPropertyUtils.visitProperties(TaskPropertyUtils.java:35)
	at org.gradle.api.internal.tasks.DefaultTaskRequiredServices.visitServiceReferences(DefaultTaskRequiredServices.java:88)
	at org.gradle.api.internal.tasks.DefaultTaskRequiredServices.collectRequiredServices(DefaultTaskRequiredServices.java:73)
	at org.gradle.api.internal.tasks.DefaultTaskRequiredServices.getElements(DefaultTaskRequiredServices.java:60)
	at org.gradle.api.internal.AbstractTask.getSharedResources(AbstractTask.java:1040)
	at org.gradle.execution.plan.LocalTaskNode.getResourcesToLock(LocalTaskNode.java:92)
	at org.gradle.execution.plan.DefaultFinalizedExecutionPlan.tryLockSharedResourceFor(DefaultFinalizedExecutionPlan.java:388)
	at org.gradle.execution.plan.DefaultFinalizedExecutionPlan.tryAcquireLocksForNode(DefaultFinalizedExecutionPlan.java:341)
	at org.gradle.execution.plan.DefaultFinalizedExecutionPlan.attemptToStart(DefaultFinalizedExecutionPlan.java:312)
	at org.gradle.execution.plan.DefaultFinalizedExecutionPlan.selectNext(DefaultFinalizedExecutionPlan.java:279)
	at org.gradle.execution.plan.DefaultPlanExecutor$MergedQueues.selectNext(DefaultPlanExecutor.java:250)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$getNextItem$1(DefaultPlanExecutor.java:433)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$$Lambda$1571/0x0000000800bd81f8.transform(Unknown Source)
	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:105)
	- locked <0x00000006306711f0> (a java.lang.Object)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.getNextItem(DefaultPlanExecutor.java:403)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:374)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.6/ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.6/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17.0.6/Thread.java:833)
```

```
 "Build operations" #407 prio=5 os_prio=31 cpu=28.45ms elapsed=118.22s tid=0x0000000141d44e00 nid=0x1450b waiting for monitor entry  [0x0000000176cfe000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at java.lang.Object.wait(java.base@17.0.6/Native Method)
	- waiting on <no object reference available>
	at java.lang.Object.wait(java.base@17.0.6/Object.java:338)
	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:113)
	- locked <0x00000006306711f0> (a java.lang.Object)
	at org.gradle.internal.work.DefaultWorkerLeaseService.acquireLocks(DefaultWorkerLeaseService.java:268)
	at org.gradle.internal.work.DefaultWorkerLeaseService.releaseWorkerLeaseAndWaitFor(DefaultWorkerLeaseService.java:385)
	at org.gradle.internal.work.DefaultWorkerLeaseService.acquireLocksWithoutWorkerLeaseWhileBlocked(DefaultWorkerLeaseService.java:373)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:247)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:109)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.runBatch(DefaultBuildOperationQueue.java:223)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.run(DefaultBuildOperationQueue.java:191)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.6/ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.6/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17.0.6/Thread.java:833)

```

**Comment by NikolayMetchev** [comment]
Discussion in the slack channel: https://gradle-community.slack.com/archives/CAHSN3LDN/p1678284115601439

**Comment by ljacomet** [comment]
this looks related to
* #23923

**Comment by NikolayMetchev** [comment]
Possibly, just to let you know that I am not using composite build.

**Comment by ljacomet** [comment]
My bad, went too quick over the thread dump. It is indeed a different issue.

**Comment by NikolayMetchev** [comment]
Sounds promising. Thank you. Hopefully a fix is easily possible. If you need me to test a beta version let me know.

**Comment by abstratt** [comment]
@NikolayMetchev @davidburstromspotify thanks, that would be helpful. We should see a 8.1 RC1 coming out today/tonight which will contain a fix for this issue. 8.2 nightlies should be getting it possibly starting tonight (the fix is in for the next release candidate, but has not made it to master yet).

**Comment by davidburstromspotify** [comment]
Thanks! I'll try it out on the 8.1 RC1. Though, as you know, due to its stochastic nature I won't know for sure if it works, only if it doesn't work. :)

**Comment by abstratt** [comment]
Just a heads-up that 8.1 RC1 has been rescheduled to Monday.

**Comment by NikolayMetchev** [comment]
Thanks, will test it out when it's ready

**Comment by abstratt** [comment]
8.1 RC1 is now available: https://docs.gradle.org/8.1-rc-1/release-notes.html

Please let me know if you still the problem with that or a later release.

**Comment by NikolayMetchev** [comment]
I can no longer reproduce the issue. It was relatively easy before.

**Comment by davidburstromspotify** [comment]
There have been other issues with RC1 that has prohibited me from building the project and provoke the issue. I'll open a new issue if it happens again, eventually.
